txnsync/txngroups.go

```diff
@@ -0,0 +1,218 @@
+	b := getMessageBuffer()
+	if cap(b) < len(uncompressedData) {
+		releaseMessageBuffer(b)
+		b = make([]byte, 0, len(uncompressedData))
+	}
+	_, compressedData, err := compress.Compress(uncompressedData, b, 1)
+	if err != nil {
+		if errors.Is(err, compress.ErrShortBuffer) {
+			s.log.Debugf("compression had negative effect, made message bigger: original msg length: %d", len(uncompressedData))
+		} else {
+			s.log.Warnf("failed to compress %d bytes txnsync msg: %v", len(uncompressedData), err)
+		}
+		releaseMessageBuffer(b)
+		return uncompressedData, compressionFormatNone
+	}
+	if len(uncompressedData) > len(compressedData)*maxCompressionRatio {
+		s.log.Infof("compression exceeded compression ratio: compressed data len: %d", len(compressedData))
+		releaseMessageBuffer(b)
+		return uncompressedData, compressionFormatNone
+	} else if len(uncompressedData) <= len(compressedData) {
+		// compression wasn't effective and ended up spending more data.
+		releaseMessageBuffer(b)
+		return uncompressedData, compressionFormatNone
+	}
+	return compressedData, compressionFormatDeflate
+}
+
+func decodeTransactionGroups(ptg packedTransactionGroups, genesisID string, genesisHash crypto.Digest) (txnGroups []transactions.SignedTxGroup, err error) {
+	data := ptg.Bytes
+	if len(data) == 0 {
+		return nil, nil
+	}
+
+	switch ptg.CompressionFormat {
+	case compressionFormatNone:
+	case compressionFormatDeflate:
+		data, err = decompressTransactionGroupsBytes(data, ptg.LenDecompressedBytes)
+		if err != nil {
+			return
+		}
+		defer releaseMessageBuffer(data)
+	default:
+		return nil, fmt.Errorf("invalid compressionFormat, %d", ptg.CompressionFormat)
+	}
+	var stub txGroupsEncodingStub
+	_, err = stub.UnmarshalMsg(data)
+	if err != nil {
+		return nil, err
+	}
+
+	if stub.TransactionGroupCount > maxEncodedTransactionGroup {
+		return nil, errors.New("invalid TransactionGroupCount")
+	}
+
+	stx := make([]transactions.SignedTxn, stub.TotalTransactionsCount)
+
+	err = stub.reconstructSignedTransactions(stx, genesisID, genesisHash)
+	if err != nil {
+		return nil, err
+	}
+
+	txnGroups = make([]transactions.SignedTxGroup, stub.TransactionGroupCount)
+	for txnCounter, txnGroupIndex := 0, 0; txnCounter < int(stub.TotalTransactionsCount); txnGroupIndex++ {
+		size := 1
+		if txnGroupIndex < len(stub.TransactionGroupSizes)*2 {
+			nibble, err := getNibble(stub.TransactionGroupSizes, txnGroupIndex)
+			if err != nil {
+				return nil, err
+			}
+			size = int(nibble) + 1
+		}
+		txnGroups[txnGroupIndex].Transactions = stx[txnCounter : txnCounter+size]
+		txnCounter += size
+	}
+
+	err = addGroupHashes(txnGroups, int(stub.TotalTransactionsCount), stub.BitmaskGroup)
+	if err != nil {
+		return nil, err
+	}
+
+	return txnGroups, nil
+}
+
+func decompressTransactionGroupsBytes(data []byte, lenDecompressedBytes uint64) (decoded []byte, err error) {
+	compressionRatio := lenDecompressedBytes / uint64(len(data)) // data should have been compressed between 0 and 95%
```

Comment:
no, division by zero is not possible, as it's being handled by the called explicitly in `decodeTransactionGroups`.

Reply:
I would still replace division with multiplication. Still, why is this check necessary?

Reply:
It is necessary because we do not want to send a compressed version if the compressed version is the same size as or bigger than the uncompressed version

Reply:
The decompressed buffer size is being allocated ahead of time, so that the decompressor won't need to perform any memory allocations. we have two "tiers" of validation -
1. make sure the the expected decoded size ( which was given to us by the other party ), is not too large.
2. make sure that the decompression ratio makes sense. ( too good, or too bad are likely to be hostile messages, and we don't want to spend the time/memory decoding these ).

Reply:
What if the compressed message is actually 20 times smaller? Do we fall back to sending the uncompressed message then?

Reply:
exactly.